src/operator/warp.cc

```diff
@@ -0,0 +1,277 @@
+/*!
+ * Copyright (c) 2016 by Contributors
+ * \file warp.cc
+ * \brief warp op
+ * \author Xu Dong
+*/
+#include "./warp-inl.h"
+#include "./mshadow_op.h"
+#include "stdlib.h"
+
+namespace mshadow {
+template<typename Dtype>
+void WarpForward(
+      const Tensor<cpu, 4, Dtype> &data,
+      const Tensor<cpu, 4, Dtype> &grid,
+      const Tensor<cpu, 4, Dtype> &out) {
+  /* assume BHWD */
+  int batchsize = data.size(0);
+  int inputImages_height = data.size(1);
+  int inputImages_width = data.size(2);
+  int output_height = out.size(1);
+  int output_width = out.size(2);
+  int inputImages_channels = data.size(3);
+
+  int output_strideWidth = out.size(3);
+  int output_strideHeight = output_strideWidth*out.size(2);
+  int output_strideBatch = output_strideHeight*out.size(1);
+
+  int inputImages_strideWidth = data.size(3);
+  int inputImages_strideHeight = inputImages_strideWidth*data.size(2);
+  int inputImages_strideBatch = inputImages_strideHeight*data.size(1);
+
+  int grids_strideWidth = grid.size(3);
+  int grids_strideHeight = grids_strideWidth*grid.size(2);
+  int grids_strideBatch = grids_strideHeight*grid.size(1);
+
+  Dtype *inputImages_data, *output_data, *grids_data;
+  inputImages_data = data.dptr_;
+  output_data = out.dptr_;
+  grids_data = grid.dptr_;
+
+  int b, yOut, xOut;
+
+  for (b=0; b < batchsize; b++) {
+    for (yOut=0; yOut < output_height; yOut++) {
+      for (xOut=0; xOut < output_width; xOut++) {
+        // read the grid
+        double xf = grids_data[b*grids_strideBatch +
+         yOut*grids_strideHeight + xOut*grids_strideWidth];
+        double yf = grids_data[b*grids_strideBatch +
+         yOut*grids_strideHeight + xOut*grids_strideWidth + 1];
+
+        // get the weights for interpolation
+        int yInTopLeft, xInTopLeft;
+        double yWeightTopLeft, xWeightTopLeft;
+
+        double xcoord = xf + xOut;
+        xInTopLeft = floor(xcoord);
+        xWeightTopLeft = std::max(1 - fabs(xcoord - xInTopLeft), 0.0);
+
+        double ycoord = yf + yOut;

```

Comment:
DType isn't good either, since coordinates should be floating point.
I think there should be two types, DataType and GridType. by default we use float for both